File: frontend/src/api/orgs.ts
import { api } from "./client";
import type { ProviderCredential } from "./providerCredentialTypes";
import type { WorkspaceListItem } from "./types";

export type OrgStatus = "pending_activation" | "active" | "suspended";
export type OrgPlan = "free" | "team" | "business" | "enterprise";
export type OrgSubscriptionStatus =
  | "inactive"
  | "active"
  | "trialing"
  | "past_due"
  | "canceled"
  | "unpaid";

export interface Organization {
  id: string;
  name: string;
  slug: string;
  createdBy: string;
  createdAt: string;
  updatedAt: string;
  status: OrgStatus;
  planId: OrgPlan;
  subscriptionStatus: OrgSubscriptionStatus;
}

export interface OrgResponse extends Organization {
  userRole?: "admin" | "member";
  memberCount: number;
}

export type CreateOrgResponse = OrgResponse;

export interface OrgMember {
  orgId: string;
  userId: string;
  username: string;
  email: string;
  role: "admin" | "member";
  /**
   * Mirrors users.email_verified from the backend. Surfaced so org admins can
   * see which members have not completed the email-verification flow and use
   * the "Verify" action to bypass it (see verifyMember).
   */
  emailVerified: boolean;
  createdAt: string;
}

export interface OrgCredential extends ProviderCredential {
  orgId: string;
  modelAllowlist: string[];
  modelContextLimits: Record<string, number>;
  /** Present when the credential was created but org-workspace auto-bind failed. */
  bindWarning?: string;
}

export interface OrgInvitation {
  id: string;
  orgId: string;
  email: string;
  role: "admin" | "member";
  invitedBy: string;
  expiresAt: string;
  acceptedAt?: string;
  declinedAt?: string;
  bounceType?: string;
  bouncedAt?: string;
  createdAt: string;
}

export interface AuditEntry {
  id: number;
  actorId: string;
  domain: string;
  action: string;
  targetId?: string;
  orgId?: string;
  metadata?: Record<string, unknown>;
  createdAt: string;
}

export interface InvitationDetail {
  orgName: string;
  orgSlug: string;
  inviterName: string;
  role: "admin" | "member";
  expiresAt: string;
}

export interface CreateOrgRequest {
  name: string;
  slug: string;
  /** Owner's email; resolved to a user ID server-side (design 0031 D1). */
  ownerEmail: string;
  planId?: OrgPlan;
}

export interface CreateInvitationsRequest {
  emails: string[];
  role: "admin" | "member";
}

export const orgsApi = {
  list: () => api.get<OrgResponse[]>("/orgs"),
  create: (req: CreateOrgRequest) =>
    api.post<CreateOrgResponse>("/orgs", req),
  get: (id: string) => api.get<OrgResponse>(`/orgs/${id}`),
  update: (id: string, req: { name?: string; slug?: string }) =>
    api.put<Organization>(`/orgs/${id}`, req),
  delete: (id: string) => api.delete<void>(`/orgs/${id}`),

  listMembers: (id: string) => api.get<OrgMember[]>(`/orgs/${id}/members`),
  addMember: (id: string, req: { userId: string; role: "admin" | "member" }) =>
    api.post<OrgMember>(`/orgs/${id}/members`, req),
  removeMember: (id: string, userId: string) =>
    api.delete<void>(`/orgs/${id}/members/${userId}`),
  changeMemberRole: (id: string, userId: string, role: "admin" | "member") =>
    api.put<{ message: string }>(`/orgs/${id}/members/${userId}`, { role }),
  /**
   * Force-verify a member's email, bypassing the email-validation flow.
   * Org-admin only. Idempotent. Used by the "Verify" button in the org admin
   * members table when an admin has confirmed the member's identity
   * out-of-band.
   */
  verifyMember: (id: string, userId: string) =>
    api.post<{ message: string }>(`/orgs/${id}/members/${userId}/verify`),

  listInvitations: (id: string) =>
    api.get<OrgInvitation[]>(`/orgs/${id}/invitations`),
  createInvitations: (id: string, req: CreateInvitationsRequest) =>
    api.post<OrgInvitation[]>(`/orgs/${id}/invitations`, req),
  revokeInvitation: (id: string, invId: string) =>
    api.delete<void>(`/orgs/${id}/invitations/${invId}`),
  resendInvitation: (id: string, invId: string) =>
    api.post<OrgInvitation>(`/orgs/${id}/invitations/${invId}/resend`),
  /**
   * Force-verify the user account associated with a pending invitation.
   * Org-admin only. Idempotent. Used when the invitee already has an
   * existing platform account but never completed email verification —
   * the admin override sets users.email_verified=true so the invitee
   * can log in. The invitation itself stays pending; the user must
   * still click the invitation link to accept and join the org.
   *
   * On 422 with body `{"error":"no_account_for_email"}`, the invitee
   * has no users row yet and must sign up before this action can be
   * used. The frontend renders a clear message in that case.
   */
  verifyInvitee: (id: string, invId: string) =>
    api.post<{ message: string }>(
      `/orgs/${id}/invitations/${invId}/verify-user`,
    ),

  getInvitationByToken: (token: string) =>
    api.get<InvitationDetail>(`/invitations/${token}`),
  acceptInvitation: (token: string) =>
    api.post<{ membership: OrgMember }>(`/invitations/${token}/accept`),
  declineInvitation: (token: string) =>
    api.post<{ status: string }>(`/invitations/${token}/decline`),

  listCredentials: (id: string) =>
    api.get<OrgCredential[]>(`/orgs/${id}/credentials`),
  createCredential: (
    id: string,
    req: {
      name: string;
      provider: string;
      apiKey: string;
      baseURL?: string;
      modelAllowlist?: string[];
      modelContextLimits?: Record<string, number>;
    },
  ) => api.post<OrgCredential>(`/orgs/${id}/credentials`, req),
  updateCredential: (
    id: string,
    credId: string,
    req: {
      name?: string;
      apiKey?: string;
      baseURL?: string;
      modelAllowlist?: string[];
      modelContextLimits?: Record<string, number>;
    },
  ) => api.put<OrgCredential>(`/orgs/${id}/credentials/${credId}`, req),
  deleteCredential: (id: string, credId: string) =>
    api.delete<void>(`/orgs/${id}/credentials/${credId}`),
  probeCredentialModels: (id: string, credId: string) =>
    api.get<{ models: { id: string; contextLimit: number }[]; baseURL?: string; warning?: string }>(
      `/orgs/${id}/credentials/${credId}/models`,
    ),

  listWorkspaces: (id: string) =>
    api.get<{ items: WorkspaceListItem[] }>(`/orgs/${id}/workspaces`),

  checkout: (id: string, planId: string) =>
    api.post<{ url: string }>(`/orgs/${id}/billing/checkout`, { planId }),
  portal: (id: string) =>
    api.post<{ url: string }>(`/orgs/${id}/billing/portal`),
  listAudit: (id: string) =>
    api.get<{ items: AuditEntry[] }>(`/orgs/${id}/audit`),
};

export interface AuditListFilters {
  orgId?: string;
  actorId?: string;
  domain?: string;
  limit?: number;
  offset?: number;
}

export interface AuditListResponse {
  items: AuditEntry[];
  pagination?: {
    total: number;
    start: number;
    end: number;
    limit: number;
    offset: number;
  };
}

export const adminAuditApi = {
  list: (filters: AuditListFilters = {}) => {
    const params = new URLSearchParams();
    if (filters.orgId) params.set("org_id", filters.orgId);
    if (filters.actorId) params.set("actor_id", filters.actorId);
    if (filters.domain) params.set("domain", filters.domain);
    if (filters.limit != null) params.set("limit", String(filters.limit));
    if (filters.offset != null) params.set("offset", String(filters.offset));
    const query = params.toString();
    return api.get<AuditListResponse>(
      `/admin/audit${query ? `?${query}` : ""}`,
    );
  },
};

// --- US-43.18: Platform admin dashboard ---

export type UserStatus = "active" | "suspended";

export interface OrgSummary extends Organization {
  memberCount: number;
  workspaceCount: number;
}

export interface UserListEntry {
  id: string;
  email: string;
  role: string;
  status: UserStatus;
  createdAt: string;
  orgCount: number;
  orgId?: string;
  orgName?: string;
}

export interface AdminListFilters {
  limit?: number;
  offset?: number;
  status?: OrgStatus | UserStatus;
}

export interface AdminListResponse<T> {
  items: T[];
  pagination?: {
    total: number;
    start: number;
    end: number;
    limit: number;
    offset: number;
  };
}

function adminListQuery(filters: AdminListFilters): string {
  const params = new URLSearchParams();
  if (filters.limit != null) params.set("limit", String(filters.limit));
  if (filters.offset != null) params.set("offset", String(filters.offset));
  if (filters.status) params.set("status", filters.status);
  const query = params.toString();
  return query ? `?${query}` : "";
}

export const adminPlatformApi = {
  listOrgs: (filters: AdminListFilters = {}) =>
    api.get<AdminListResponse<OrgSummary>>(
      `/admin/orgs${adminListQuery(filters)}`,
    ),
  listUsers: (filters: AdminListFilters = {}) =>
    api.get<AdminListResponse<UserListEntry>>(
      `/admin/users${adminListQuery(filters)}`,
    ),
  suspendOrg: (orgId: string) =>
    api.post<{ status: string }>(`/admin/orgs/${orgId}/suspend`),
  unsuspendOrg: (orgId: string) =>
    api.post<{ status: string }>(`/admin/orgs/${orgId}/unsuspend`),
  suspendUser: (userId: string, force = false) =>
    api.post<{ status: string }>(
      `/admin/users/${userId}/suspend${force ? "?force=true" : ""}`,
    ),
  unsuspendUser: (userId: string) =>
    api.post<{ status: string }>(`/admin/users/${userId}/unsuspend`),
};

File: pkg/types/orgs.go
// Copyright (C) 2026 Michael Kao
// SPDX-License-Identifier: AGPL-3.0-or-later

package types

import "time"

// OrgRole represents a user's role within an organization.
type OrgRole string

const (
	OrgRoleAdmin  OrgRole = "admin"
	OrgRoleMember OrgRole = "member"
)

// OrgStatus is the operational status of an organization. It gates access:
// only non-suspended orgs are usable via OrgMemberGuard/OrgAdminGuard. Both
// 'active' and 'pending_activation' allow access (the creator needs to reach
// the portal and Stripe checkout while pending); 'suspended' is fully locked.
type OrgStatus string

const (
	OrgStatusPendingActivation OrgStatus = "pending_activation"
	OrgStatusActive            OrgStatus = "active"
	OrgStatusSuspended         OrgStatus = "suspended"
)

// OrgPlan is the product plan identifier stored in organizations.plan_id and
// used locally for feature gating. The plan is set at org creation
// (enterprise for platform-admin orgs; the selected checkout plan on
// checkout.session.completed for self-service orgs). Per-event plan syncing
// from Stripe is planned for US-43.15.
type OrgPlan string

const (
	PlanFree       OrgPlan = "free"
	PlanTeam       OrgPlan = "team"
	PlanBusiness   OrgPlan = "business"
	PlanEnterprise OrgPlan = "enterprise"
)

// OrgSubscriptionStatus tracks the Stripe subscription lifecycle separately
// from OrgStatus. An org can be status='active' (members retain access) while
// subscription_status='past_due' (in the 7-day Smart Retries grace window).
type OrgSubscriptionStatus string

const (
	SubscriptionInactive OrgSubscriptionStatus = "inactive"
	SubscriptionActive   OrgSubscriptionStatus = "active"
	SubscriptionTrialing OrgSubscriptionStatus = "trialing"
	SubscriptionPastDue  OrgSubscriptionStatus = "past_due"
	SubscriptionCanceled OrgSubscriptionStatus = "canceled"
	SubscriptionUnpaid   OrgSubscriptionStatus = "unpaid"
)

// Organization is the API DTO for an organization.
type Organization struct {
	ID                 string                `json:"id"`
	Name               string                `json:"name"`
	Slug               string                `json:"slug"`
	CreatedBy          string                `json:"createdBy"`
	CreatedAt          time.Time             `json:"createdAt"`
	UpdatedAt          time.Time             `json:"updatedAt"`
	Status             OrgStatus             `json:"status"`
	PlanID             OrgPlan               `json:"planId"`
	SubscriptionStatus OrgSubscriptionStatus `json:"subscriptionStatus"`
}

// OrgMember is the API DTO for an organization membership.
//
// EmailVerified mirrors users.email_verified for the member's user account. It
// is exposed so org admins can see which members have not yet completed the
// email-verification flow, and offers a "Verify" action to bypass it
// (POST /orgs/:id/members/:userID/verify). Verification state lives on the
// user row (not the membership) — there is exactly one user account per
// member, and a single user belongs to at most one org under single-org
// enforcement (D8).
type OrgMember struct {
	OrgID         string    `json:"orgId"`
	UserID        string    `json:"userId"`
	Username      string    `json:"username"`
	Email         string    `json:"email"`
	Role          OrgRole   `json:"role"`
	EmailVerified bool      `json:"emailVerified"`
	CreatedAt     time.Time `json:"createdAt"`
}

// CreateOrgRequest is the request body for creating an organization. The slug is
// lowercased by the service before insert and uniqueness check.
//
// Per design 0031 D1, org creation is platform-admin only. The admin supplies
// the intended owner's email; the backend resolves it to a user ID. This is a
// single lookup, not a search/list endpoint (account-enumeration prevention).
//
// Slug format: lowercase letters, digits, and single hyphens between segments
// (e.g. "my-org", "team-1"). The `slug` validator is registered in
// pkg/types/validators.go. Hyphens are required because the frontend's
// slugify() produces them from multi-word names; rejecting hyphens would
// produce an unreachable 400 for any user-friendly name.
type CreateOrgRequest struct {
	Name       string  `json:"name"       binding:"required,min=2,max=100"`
	Slug       string  `json:"slug"       binding:"required,min=2,max=50,slug"`
	OwnerEmail string  `json:"ownerEmail" binding:"required,email"`
	PlanID     OrgPlan `json:"planId"     binding:"omitempty"`
}

// UpdateOrgRequest is the request body for updating an organization.
type UpdateOrgRequest struct {
	Name string `json:"name" binding:"omitempty,min=2,max=100"`
	Slug string `json:"slug" binding:"omitempty,min=2,max=50,slug"`
}

// CreateOrgResponse is returned by POST /api/v1/orgs. Org creation is
// platform-admin only (design 0031 D1).
type CreateOrgResponse struct {
	OrgResponse
}

// OrgResponse extends Organization with the calling user's membership context.
// UserRole is omitempty so that an empty string (caller is not a member — e.g.
// platform admin creating an org for someone else) is omitted from JSON rather
// than appearing as `"userRole": ""`.
type OrgResponse struct {
	Organization
	UserRole    OrgRole `json:"userRole,omitempty"`
	MemberCount int     `json:"memberCount"`
}

// AddOrgMemberRequest is the request body for adding an org member.
type AddOrgMemberRequest struct {
	UserID string  `json:"userId" binding:"required"`
	Role   OrgRole `json:"role"   binding:"required"`
}

// ChangeOrgMemberRoleRequest is the request body for changing a member's role.
type ChangeOrgMemberRoleRequest struct {
	Role OrgRole `json:"role" binding:"required"`
}

// --- US-43.2: Org invitations ---

// OrgInvitation is the API DTO for an org invitation row.
type OrgInvitation struct {
	ID         string     `json:"id"`
	OrgID      string     `json:"orgId"`
	Email      string     `json:"email"`
	Role       OrgRole    `json:"role"`
	InvitedBy  string     `json:"invitedBy"`
	TokenHash  string     `json:"-"`
	ExpiresAt  time.Time  `json:"expiresAt"`
	AcceptedAt *time.Time `json:"acceptedAt,omitempty"`
	DeclinedAt *time.Time `json:"declinedAt,omitempty"`
	BounceType string     `json:"bounceType,omitempty"`
	BouncedAt  *time.Time `json:"bouncedAt,omitempty"`
	CreatedAt  time.Time  `json:"createdAt"`

	// InviteeUserExists is true when a `users` row exists with this
	// invitation's email (case-folded match). Surfaced so the org admin
	// UI can render the per-row Verify button only when force-verify is
	// actionable. Pointer so a missing-from-payload value is
	// distinguishable from a definite false on older API responses.
	InviteeUserExists *bool `json:"inviteeUserExists,omitempty"`

	// InviteeEmailVerified mirrors users.email_verified for the row
	// matched by InviteeUserExists. Nil when no users row exists. The
	// org admin UI hides the Verify button when this is true (the
	// override has already been applied or the user verified through
	// the normal flow).
	InviteeEmailVerified *bool `json:"inviteeEmailVerified,omitempty"`
}

// CreateInvitationsRequest is the body for POST /orgs/:id/invitations.
type CreateInvitationsRequest struct {
	Emails []string `json:"emails" binding:"required,min=1,max=100,dive,email"`
	Role   OrgRole  `json:"role"   binding:"required"`
}

// InvitationDetail is the public response for GET /invitations/:token. It does
// not expose the token hash or internal IDs beyond what the recipient needs to
// decide whether to accept.
type InvitationDetail struct {
	OrgName     string    `json:"orgName"`
	OrgSlug     string    `json:"orgSlug"`
	InviterName string    `json:"inviterName"`
	Role        OrgRole   `json:"role"`
	ExpiresAt   time.Time `json:"expiresAt"`
}

// LastAdminOrg identifies an org where a user is the sole active admin.
type LastAdminOrg struct {
	OrgID   string `json:"orgId"`
	OrgName string `json:"orgName"`
}

// OrgSummary extends Organization with aggregate counts for the platform
// admin dashboard. The counts are populated by a single SQL query (no N+1).
type OrgSummary struct {
	Organization
	MemberCount    int `json:"memberCount"`
	WorkspaceCount int `json:"workspaceCount"`
}

// OrgSSOConfig is the per-org OIDC SSO configuration. ClientSecret is the
// encrypted blob stored in the DB; it is never serialized to JSON.
// VerifiedDomains is a subset of ClaimedDomains that have passed DNS
// verification (D17 Q-S2); only verified domains auto-route on the login
// page. VerificationToken is the per-org random token the org admin places
// as a TXT record at _llmsafespaces-verify.<domain> to prove ownership.
type OrgSSOConfig struct {
	OrgID             string             `json:"-"`
	DiscoveryURL      string             `json:"discoveryUrl"`
	ClientID          string             `json:"clientId"`
	ClientSecret      []byte             `json:"-"`
	ClaimedDomains    []string           `json:"claimedDomains"`
	VerifiedDomains   []string           `json:"verifiedDomains"`
	VerificationToken string             `json:"verificationToken"`
	AutoProvision     bool               `json:"autoProvision"`
	GroupRoleMapping  map[string]OrgRole `json:"groupRoleMapping"`
	CreatedAt         time.Time          `json:"createdAt"`
	UpdatedAt         time.Time          `json:"updatedAt"`
}

// OrgSSOConfigResponse is the API response shape — omits the encrypted secret.
type OrgSSOConfigResponse struct {
	OrgID             string             `json:"orgId"`
	DiscoveryURL      string             `json:"discoveryUrl"`
	ClientID          string             `json:"clientId"`
	HasSecret         bool               `json:"hasSecret"`
	ClaimedDomains    []string           `json:"claimedDomains"`
	VerifiedDomains   []string           `json:"verifiedDomains"`
	VerificationToken string             `json:"verificationToken"`
	AutoProvision     bool               `json:"autoProvision"`
	GroupRoleMapping  map[string]OrgRole `json:"groupRoleMapping"`
	UpdatedAt         time.Time          `json:"updatedAt"`
}

// UpsertSSOConfigRequest is the API request body for creating/updating SSO config.
type UpsertSSOConfigRequest struct {
	DiscoveryURL     string             `json:"discoveryUrl"     binding:"required,url"`
	ClientID         string             `json:"clientId"         binding:"required,min=1"`
	ClientSecret     string             `json:"clientSecret"     binding:"omitempty"`
	ClaimedDomains   []string           `json:"claimedDomains"`
	AutoProvision    *bool              `json:"autoProvision"`
	GroupRoleMapping map[string]OrgRole `json:"groupRoleMapping"`
}

// SSODomain is a single entry in the domain discovery response.
type SSODomain struct {
	Domain  string `json:"domain"`
	OrgSlug string `json:"orgSlug"`
	OrgName string `json:"orgName"`
}
